### ISSUE TITLE
Unhelpful error message when using program equality in logic

### ISSUE THREAD

**Issue by shiatsumat** [xldenis/creusot#97]
Currently, the test `specification/logic_call` has the following.
```rust
logic!{
	fn reflexive<T: PartialEq>(x: T) -> bool {
		true == true
	}
}
```
This is awkward. Clearly, what we expect here, especially by the name `reflexive`, is the following.
```rust
logic!{
	fn reflexive<T: PartialEq>(x: T) -> bool {
		x == x
	}
}
```
But actually this doesn't work.

Creusot outputs the following MLCFG code for `PartialEq` and the `reflexive` function.
```ocaml
...
module Core_Cmp_PartialEq
  type self   
  type rhs   
  use prelude.Prelude
  val eq (self : self) (other : rhs) : bool
  val ne (self : self) (other : rhs) : bool
end
module LogicCall_Reflexive_Interface
  type t   
  function reflexive (x : t) : bool
end
module LogicCall_Reflexive
  type t   
  clone Core_Cmp_PartialEq as PartialEq0 with type self = t, type rhs = t
  function reflexive (x : t) : bool = 
    PartialEq0.eq x x
end
...
```
And Why3 claims that `PartialEq0.eq` is not defined yet, reporting like the following.
```
File "logic_call.stdout", line x, characters x-x:
unbound function or predicate symbol 'PartialEq0.eq'
```

**Comment by xldenis** [comment]
That's normal, the error could be improved but the result is expected. `eq` is a _program_ symbol, not a logical one, you cannot use it in logic. Instead you should use `equal` for _logical equality_ between any types.

**Comment by xldenis** [comment]
However, I actually opened a related issue in Why3's issue tracker for it to give a better error message. https://gitlab.inria.fr/why3/why3/-/issues/591

**Comment by shiatsumat** [comment]
OK, I see.

**Comment by shiatsumat** [comment]
But what's the point of the `reflexive` test then? Should this be fixed?

**Comment by xldenis** [comment]
you're right, it should be changed to `equal(x, x)` with the trait bound removed. I think the idea was that in the future the `Eq` trait will have a logical component representing the _logical_ equality being realized and that would be used in `reflexive`. Change it for now.

**Comment by shiatsumat** [comment]
Well, actually Rust's `PartialEq::eq` is not logically equal to `equal(x, y)` (e.g., `eq` on `&mut T` just compares the current target values, not the addresses).
Only for some known primitive types like `u32`, `PartialEq::eq` is logically equal to `equal`.

I suppose the `logic_call` test is meant to check the type checking performed over logic functions, so probably we should modify the test?

**Comment by xldenis** [comment]
Yes thats what I mean, we would only require `eq` to implement _some_ equivalence relation, not necessarily `=`. In my head the trait would look like:

```rust
trait Eq {
  logic! { fn equiv_rel(&self, &self) -> bool; }

  #[ensures(result <-> equiv_rel(self, o))]
  fn eq(&self, o : &self) -> bool;

  law equiv_refl : ...
  law equiv_trans : ...
  law equiv_sym : ... 
}
```

**Comment by xldenis** [comment]
Can you modify the test and submit a PR?

**Comment by shiatsumat** [comment]
Well, actually I'm not sure how the test should be modified... Could you modify it instead?

**Comment by shiatsumat** [comment]
Note: the test `logic_call` was fixed by #99.

**Comment by xldenis** [comment]
We now generate an error specifically saying you attempted to call a program function in the logic (and the otherway around).
